framework/src/meshmodifiers/BreakMeshByBlock.C

```diff
@@ -50,91 +50,96 @@ BreakMeshByBlock::modify()
 
     if (current_node != nullptr)
     {
-
       // find node multiplicity
       const std::set<SubdomainID> & connected_blocks = _mesh_ptr->getNodeBlockIds(*current_node);
       unsigned int node_multiplicity = connected_blocks.size();
 
       // check if current_node need to be duplicated
       if (node_multiplicity > 1)
       {
-
         // retrieve connected elements from the map
         const std::vector<dof_id_type> & connected_elems = node_it->second;
 
         // find reference_subdomain_id (e.g. the subdomain with lower id)
         auto subdomain_it = connected_blocks.begin();
         SubdomainID reference_subdomain_id = *subdomain_it;
 
-        // we need to keep track of which element belong to the reference block and which
-        // to neighboring blocks
-        std::set<dof_id_type> reference_elems;
-        std::set<dof_id_type> neighbor_elems;
-
-        // initialize current multiplicity to keep track of how many nodes we added
-        unsigned int current_multiplicity = 1;
-
-        // loop over node multiplicity to add nodes starting from the second subdomain
-        // the original node remains on the reference_subdoamin
-        while (current_multiplicity < node_multiplicity)
+        // multiplicity counter to keep track of how many nodes we added
+        unsigned int multiplicity_counter = node_multiplicity;
+        for (auto elem_id : connected_elems)
         {
+          // all the duplicate nodes are added and assigned
+          if (multiplicity_counter == 0)
+            break;
 
-          // add new node
-          Node * new_node = Node::build(*current_node, mesh.n_nodes()).release();
-          new_node->processor_id() = current_node->processor_id();
-          mesh.add_node(new_node);
-          current_multiplicity += 1; // node created, update multiplicity
-
-          // fix all the elements belonging to the connected_blocks[subdomain_it]
-          for (auto elem_id : connected_elems)
+          Elem * current_elem = mesh.elem_ptr(elem_id);
+          if (current_elem->subdomain_id() != reference_subdomain_id)
           {
-            Elem * current_elem = mesh.elem_ptr(elem_id);
-
-            // check if we need to fix a node in the current element
-            // if current_elem belong to the reference subdomain
-            if (current_elem->subdomain_id() == reference_subdomain_id)
-              reference_elems.insert(current_elem->id()); // store reference element
-            else                                          // this is a neighbor element
+            // assign the newly added node to current_elem
+            Node * new_node = nullptr;
+
+            for (unsigned int local_elem_node_it = 0; local_elem_node_it < current_elem->n_nodes();
+                 ++local_elem_node_it)
```

Comment:
This is not the best variable name. `it` is usually short for `iterator`, but you are creating an index. It makes the use of the variable very unintuitive within the loop.